# Import bibliotek i danych

In [ ]:
import pandas as pd
import statsmodels.formula.api as smf
from statsmodels.stats.outliers_influence import variance_inflation_factor 
import statsmodels.api as sm
from sklearn.model_selection import KFold
from sklearn.metrics import *
from sklearn.linear_model import LinearRegression
data=pd.read_csv("data_fe.csv")
data=data.drop(columns=['Unnamed: 0'])

# Oszacowanie modelu OLS

In [7]:
features=data.columns.tolist()
features.remove('stars')
wzor = 'stars~' + '+'.join(features)
mod = smf.ols(formula = wzor, 
              data = data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  stars   R-squared:                       0.411
Model:                            OLS   Adj. R-squared:                  0.408
Method:                 Least Squares   F-statistic:                     123.5
Date:                Sun, 21 Apr 2024   Prob (F-statistic):               0.00
Time:                        17:33:26   Log-Likelihood:                 286.89
No. Observations:                6930   AIC:                            -493.8
Df Residuals:                    6890   BIC:                            -220.0
Df Model:                          39                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercept                            2.4971      0.104     24.059      0.000       2.294       2.701
Crime                                0.0121      0.011      1.067      0.286      -0.010       0.034
Romans                              -0.1011      0.009    -11.391      0.000      -0.118      -0.084
sad_mm                               0.2771      0.022     12.399      0.000       0.233       0.321
reviews_boxcox                      -0.3770      0.033    -11.312      0.000      -0.442      -0.312
reflective_boxcox                    0.0222      0.007      3.396      0.001       0.009       0.035
mysterious_rs                        0.0434      0.011      4.123      0.000       0.023       0.064
inspiring_log                        0.0119      0.003      3.497      0.000       0.005       0.018
informative_boxcox                   0.0028      0.001      2.395      0.017       0.001       0.005
emotional_boxcox                     0.0137      0.017      0.806      0.420      -0.020       0.047
challenging_mm                       0.4158      0.099      4.179      0.000       0.221       0.611
author_stars                         0.0031      0.003      1.096      0.273      -0.002       0.009
Fantasy_author_stars                 0.0057      0.003      1.729      0.084      -0.001       0.012
Fantasy_mix_qt                       0.0081      0.018      0.446      0.656      -0.028       0.044
Fiction_adventurous_qt               0.1284      0.020      6.573      0.000       0.090       0.167
Fiction_author_count_ss              0.0192      0.003      5.847      0.000       0.013       0.026
Fiction_challenging_mm              -0.1085      0.097     -1.114      0.265      -0.299       0.082
Fiction_character_ss                -0.0545      0.005     -9.932      0.000      -0.065      -0.044
Fiction_emotional_boxcox            -0.1167      0.020     -5.754      0.000      -0.156      -0.077
Fiction_hopeful_rs                   0.1096      0.008     14.002      0.000       0.094       0.125
Fiction_mix_qt                       0.1839      0.022      8.542      0.000       0.142       0.226
Fiction_pages_qt                    -0.0184      0.013     -1.432      0.152      -0.044       0.007
Fiction_plot_qt                     -0.3140      0.023    -13.761      0.000      -0.359      -0.269
Literary_author_stars                0.0082      0.003      2.697      0.007       0.002       0.014
Literary_emotional_boxcox           -0.0074      0.015     -0.501      0.616      -0.036       0.021
Literary_hopeful_rs                  0.0214      0.010      2.245      0.025       0.003       0.040
Literary_inspiring_log              -0.0097      0.004     -2.357      0.018      -0.018      -0.002
Literary_mysterious_rs              -0.0114      0.01

Dużo zmiennych wyszło jako nieistotne statystyczne, nawet te o dużej wartości MI, dlatego przeprowadzimy analizę VIF współliniowości

In [14]:
from statsmodels.stats.outliers_influence import variance_inflation_factor 
# the independent variables set 
X = data[features]
  
# VIF dataframe 
vif_data = pd.DataFrame() 
vif_data["feature"] = X.columns 
  
# calculating VIF for each feature 
vif_data["VIF"] = [variance_inflation_factor(X.values, i) 
                          for i in range(len(X.columns))] 

vif_data.sort_values(by=["VIF"], ascending=False)

feature        VIF
9                     challenging_mm  81.239825
15            Fiction_challenging_mm  59.824637
3                     reviews_boxcox  53.825270
17          Fiction_emotional_boxcox  23.885668
21                   Fiction_plot_qt  22.052404
8                   emotional_boxcox  19.886851
19                    Fiction_mix_qt  19.626884
29         Nonfiction_challenging_mm  18.133970
13            Fiction_adventurous_qt  15.719166
4                  reflective_boxcox  11.011548
10                      author_stars  10.082686
35                     series_mix_qt   9.221221
6                      inspiring_log   8.643754
38         lighthearted_mm__funny_qt   8.090098
12                    Fantasy_mix_qt   7.942337
34              series_inspiring_log   6.575121
20                  Fiction_pages_qt   6.432153
2                             sad_mm   6.351058
33               series_author_stars   5.735277
11              Fantasy_author_stars   5.696650
5                      mysterious_rs   5.544653
23         Literary_emotional_boxcox   4.823866
25            Literary_inspiring_log   4.764064
22             Literary_author_stars   4.173639
18                Fiction_hopeful_rs   3.479166
16              Fiction_character_ss   3.091898
36  lighthearted_mm__relaxing_boxcox   3.064561
0                              Crime   2.741696
32             Social_challenging_mm   2.714269
28           Nonfiction_author_stars   2.555901
37                  tense_rs_dark_ss   2.440615
26            Literary_mysterious_rs   2.371435
7                 informative_boxcox   2.157252
1                             Romans   2.143265
30                Nonfiction_dark_ss   2.073852
31               Social_author_stars   1.955174
24               Literary_hopeful_rs   1.844379
27        Nonfiction_author_count_ss   1.499982
14           Fiction_author_count_ss   1.280410

Na początek usuniemy z modelu zmienne: Fiction_challenging_mm, Fiction_emotional_boxcox, Fantasy_mix_qt, author_stars	

In [17]:
features.remove('Fiction_challenging_mm')
features.remove('Fiction_emotional_boxcox')
features.remove('Fantasy_mix_qt')
features.remove('author_stars')
wzor = 'stars~' + '+'.join(features)
mod = smf.ols(formula = wzor, 
              data = data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  stars   R-squared:                       0.408
Model:                            OLS   Adj. R-squared:                  0.405
Method:                 Least Squares   F-statistic:                     135.9
Date:                Sun, 21 Apr 2024   Prob (F-statistic):               0.00
Time:                        18:27:29   Log-Likelihood:                 268.52
No. Observations:                6930   AIC:                            -465.0
Df Residuals:                    6894   BIC:                            -218.7
Df Model:                          35                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercept                            2.4724      0.103     23.995      0.000       2.270       2.674
Crime                                0.0119      0.011      1.083      0.279      -0.010       0.033
Romans                              -0.0984      0.009    -11.128      0.000      -0.116      -0.081
sad_mm                               0.2938      0.022     13.250      0.000       0.250       0.337
reviews_boxcox                      -0.3904      0.033    -11.847      0.000      -0.455      -0.326
reflective_boxcox                    0.0213      0.007      3.244      0.001       0.008       0.034
mysterious_rs                        0.0369      0.010      3.542      0.000       0.016       0.057
inspiring_log                        0.0097      0.003      2.890      0.004       0.003       0.016
informative_boxcox                   0.0023      0.001      1.956      0.050   -4.53e-06       0.005
emotional_boxcox                    -0.0417      0.014     -3.031      0.002      -0.069      -0.015
challenging_mm                       0.2923      0.028     10.273      0.000       0.237       0.348
Fantasy_author_stars                 0.0078      0.002      3.291      0.001       0.003       0.012
Fiction_adventurous_qt               0.1095      0.019      5.832      0.000       0.073       0.146
Fiction_author_count_ss              0.0199      0.003      6.141      0.000       0.014       0.026
Fiction_character_ss                -0.0651      0.005    -12.641      0.000      -0.075      -0.055
Fiction_hopeful_rs                   0.1170      0.008     15.121      0.000       0.102       0.132
Fiction_mix_qt                       0.1885      0.021      9.152      0.000       0.148       0.229
Fiction_pages_qt                    -0.0155      0.013     -1.210      0.226      -0.041       0.010
Fiction_plot_qt                     -0.3751      0.020    -18.417      0.000      -0.415      -0.335
Literary_author_stars                0.0101      0.003      3.978      0.000       0.005       0.015
Literary_emotional_boxcox           -0.0211      0.014     -1.464      0.143      -0.049       0.007
Literary_hopeful_rs                  0.0219      0.009      2.319      0.020       0.003       0.040
Literary_inspiring_log              -0.0091      0.004     -2.247      0.025      -0.017      -0.001
Literary_mysterious_rs              -0.0080      0.012     -0.649      0.516      -0.032       0.016
Nonfiction_author_count_ss          -0.0688      0.013     -5.327      0.000      -0.094      -0.044
Nonfiction_author_stars              0.0224      0.004      6.386      0.000       0.016       0.029
Nonfiction_challenging_mm            0.4494      0.043     10.393      0.000       0.365       0.534
Nonfiction_dark_ss                  -0.0589      0.01

Pozostały zmienne nieistotne statystycznie: Crime, Fiction_pages_qt, series_inspiring_log, Social_challenging_mm, Literary_mysterious_rs, Literary_emotional_boxcox, które mają niską wartość VIF.

In [33]:
features.remove('Crime')
features.remove('Fiction_pages_qt')
features.remove('series_inspiring_log')
features.remove('Social_challenging_mm')
features.remove('Literary_mysterious_rs')
features.remove('Literary_emotional_boxcox')
features.remove('Literary_inspiring_log')
wzor = 'stars~' + '+'.join(features)
mod = smf.ols(formula = wzor, 
              data = data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  stars   R-squared:                       0.407
Model:                            OLS   Adj. R-squared:                  0.405
Method:                 Least Squares   F-statistic:                     169.3
Date:                Sun, 21 Apr 2024   Prob (F-statistic):               0.00
Time:                        18:46:57   Log-Likelihood:                 262.52
No. Observations:                6930   AIC:                            -467.0
Df Residuals:                    6901   BIC:                            -268.6
Df Model:                          28                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercept                            2.4795      0.103     24.171      0.000       2.278       2.681
Romans                              -0.1023      0.009    -11.735      0.000      -0.119      -0.085
sad_mm                               0.2989      0.022     13.565      0.000       0.256       0.342
reviews_boxcox                      -0.3892      0.033    -11.895      0.000      -0.453      -0.325
reflective_boxcox                    0.0197      0.006      3.076      0.002       0.007       0.032
mysterious_rs                        0.0358      0.008      4.422      0.000       0.020       0.052
inspiring_log                        0.0086      0.002      3.637      0.000       0.004       0.013
informative_boxcox                   0.0024      0.001      2.085      0.037       0.000       0.005
emotional_boxcox                    -0.0443      0.013     -3.365      0.001      -0.070      -0.018
challenging_mm                       0.2860      0.026     10.806      0.000       0.234       0.338
Fantasy_author_stars                 0.0074      0.002      3.282      0.001       0.003       0.012
Fiction_adventurous_qt               0.1002      0.018      5.521      0.000       0.065       0.136
Fiction_author_count_ss              0.0195      0.003      6.091      0.000       0.013       0.026
Fiction_character_ss                -0.0646      0.005    -12.596      0.000      -0.075      -0.055
Fiction_hopeful_rs                   0.1137      0.007     15.244      0.000       0.099       0.128
Fiction_mix_qt                       0.1919      0.019      9.862      0.000       0.154       0.230
Fiction_plot_qt                     -0.3728      0.020    -18.649      0.000      -0.412      -0.334
Literary_author_stars                0.0104      0.002      5.371      0.000       0.007       0.014
Literary_hopeful_rs                  0.0241      0.009      2.753      0.006       0.007       0.041
Nonfiction_author_count_ss          -0.0671      0.013     -5.223      0.000      -0.092      -0.042
Nonfiction_author_stars              0.0213      0.003      6.246      0.000       0.015       0.028
Nonfiction_challenging_mm            0.4732      0.039     12.117      0.000       0.397       0.550
Nonfiction_dark_ss                  -0.0594      0.010     -6.092      0.000      -0.079      -0.040
Social_author_stars                  0.0116      0.002      4.674      0.000       0.007       0.016
series_author_stars                  0.0129      0.003      4.836      0.000       0.008       0.018
series_mix_qt                        0.1040      0.015      6.747      0.000       0.074       0.134
lighthearted_mm__relaxing_boxcox    -0.0042      0.001     -2.835      0.005      -0.007      -0.001
tense_rs_dark_ss                     0.3686      0.11

Reszta zmiennych w modelu jest istotna statystycznie

Większość zmiennych wpływa pozytywnie na zmienną objaśnianą oprócz: Romans, reviews, emotional, Fiction_character, Fiction_plot,
Nonfiction_author_count, Nonfiction_dark oraz lighthearted_relaxing.

# Jakość predykcyjna modelu OLS

In [64]:
rmse=[]
kf = KFold(n_splits=10, shuffle=True, random_state=2024)
for train, test in kf.split(data.index.values): 
    mod = LinearRegression().fit(data[features].iloc[train], data.iloc[train]["stars"])
    y_pred = mod.predict(data[features].iloc[test])
    y_test=data.iloc[test]["stars"]
    rmse.append(mean_squared_error(y_test, y_pred)**0.5)
print("Linear Regression RMSE: {}".format(round(sum(rmse)/len(rmse),5)))

Linear Regression RMSE: 0.23362


Można jeszcze pokazać jak słaby to jest wynik licząc moc predykcji modelu bez feature engineeringu.

In [61]:
df=pd.read_csv("data_eda.csv")
df=df.drop(columns=['Unnamed: 0'])

ft=df.columns.tolist()
ft.remove('stars')

wzor = 'stars~' + '+'.join(ft)
mod = smf.ols(formula = wzor, 
              data = df)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  stars   R-squared:                       0.462
Model:                            OLS   Adj. R-squared:                  0.460
Method:                 Least Squares   F-statistic:                     191.3
Date:                Sun, 21 Apr 2024   Prob (F-statistic):               0.00
Time:                        19:41:20   Log-Likelihood:                 599.87
No. Observations:                6930   AIC:                            -1136.
Df Residuals:                    6898   BIC:                            -916.7
Df Model:                          31                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        3.0092      0.032     94.115      0.000       2.946       3.072
pages        -3.883e-05      2e-05     -1.940      0.052   -7.81e-05    4.07e-07
reviews       9.675e-07   1.08e-07      8.988      0.000    7.56e-07    1.18e-06
series           0.0750      0.007     10.502      0.000       0.061       0.089
mix              0.0690      0.030      2.333      0.020       0.011       0.127
character       -0.0956      0.024     -3.962      0.000      -0.143      -0.048
plot            -0.1866      0.035     -5.407      0.000      -0.254      -0.119
funny            0.4076      0.019     21.337      0.000       0.370       0.445
lighthearted    -0.0072      0.032     -0.221      0.825      -0.071       0.057
emotional        0.3704      0.025     14.836      0.000       0.321       0.419
hopeful          0.3196      0.035      9.126      0.000       0.251       0.388
inspiring        0.2427      0.033      7.459      0.000       0.179       0.306
relaxing         0.7192      0.066     10.872      0.000       0.590       0.849
tense            0.3063      0.030     10.294      0.000       0.248       0.365
sad              0.2198      0.026      8.473      0.000       0.169       0.271
reflective       0.0202      0.025      0.816      0.414      -0.028       0.069
adventurous      0.1904      0.017     11.365      0.000       0.158       0.223
challenging      0.4078      0.026     15.404      0.000       0.356       0.460
informative      0.4372      0.022     19.643      0.000       0.394       0.481
mysterious       0.1906      0.018     10.760      0.000       0.156       0.225
dark            -0.0024      0.019     -0.125      0.901      -0.040       0.036
author_count     0.0025      0.000      5.897      0.000       0.002       0.003
author_stars     0.0188      0.002     11.875      0.000       0.016       0.022
Fiction          0.0797      0.029      2.786      0.005       0.024       0.136
Nonfiction       0.1617      0.015     10.645      0.000       0.132       0.191
Literary         0.0278      0.007      3.782      0.000       0.013       0.042
Fantasy          0.0201      0.009      2.182      0.029       0.002       0.038
Crime           -0.0041      0.011     -0.371      0.711      -0.026       0.017
Social           0.0485      0.008      6.274      0.000       0.033       0.064
Children        -0.0304      0.008     -3.959      0.000      -0.045      -0.015
Romans          -0.0620      0.009     -6.794      0.000      -0.080      -0.044
Realism         -0.0504      0.008     -6.317      0.000      -0.066      -0.035
==============================================================================
Omnibus:                      238.608   Durbin-Watson:                   1.931
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              343.875
Skew:      

In [62]:
rmse=[]
kf = KFold(n_splits=10, shuffle=True, random_state=2024)
for train, test in kf.split(df.index.values):
    mod = LinearRegression().fit(df[ft].iloc[train], df.iloc[train]["stars"])
    y_pred = mod.predict(df[ft].iloc[test])
    y_test=df.iloc[test]["stars"]
    rmse.append(mean_squared_error(y_test, y_pred)**0.5)
print("Linear Regression RMSE: {}".format(round(sum(rmse)/len(rmse),5)))

Linear Regression RMSE: 0.22289


Widzimy, że w przypadku modelu regresji feature engineering wręcz pogorszył naszą moc predycyjną, jednak możemy nadal ustalić te wartości jako bazowe dla dalszych modeli: *0.23362* oraz *0.22289*